### PR TITLE
Add wsclient library to ecs-agent module

### PR DESCRIPTION
**Pull request by RichaGangwar**
### Summary
Add wsclient library to ecs-agent

### Implementation details
1. Added a new directory wsclient to ecs-agent.
2. Added the new files in wsclient related to wsclient functionality
3. Added utils directory to move the needed dependencies from agent to ecs-agent.
4. Edited the needed imports in agent related to moved utils.
5. Moved Proxy( *http.Request) function to  proxy.go from utils.go to avoid moving the whole utils.go in ecs-agent.

TODOs to be handled in future PRs:
1. Periodic timeout/disconnection of WebSocket connections in wsclient instead of tcs/acs clients
2. Move exitcodes package to /ecs-agent
3. Use context.Canceled.Error() instead of hardcoding "context canceled"
4. Replace GetMockServer() used in unit tests with gomock generated mock interfaces.
5. Remove docker specific checks in wsclient library.
6. Remove existing agent/wsclient after plugging in the new libraries in ecs-agent, i.e. tcs and acs.

### Testing
Unit tested with existing tests in wsclient
Unit tested agent with make test

New tests cover the changes: no

### Description for the changelog
Add wsclient library to ecs-agent module.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by danehlim** [comment]
Linux / Linux unit tests check is failing at the moment, just want to confirm is this PR ready for review or currently still WIP?

**Comment by RichaGangwar** [comment]
> Linux / Linux unit tests check is failing at the moment, just want to confirm is this PR ready for review or currently still WIP?
@danehlim Its ready for review. The unit tests passed in my local setup. The Linux unit tests is failing here due to less coverage than expected. Ideally, we should not see the coverage being lowered as we are moving the existing files. I am checking on it.

File: ecs-agent/wsclient/client.go
Comment (danehlim) [inline ecs-agent/wsclient/client.go:167]: This needs to be capitalized to be exported.

File: ecs-agent/wsclient/client.go
Comment (danehlim) [inline ecs-agent/wsclient/client.go:44]: I think we want to be using github.com/aws/amazon-ecs-agent/ecs-agent/logger for any logging in `ecs-agent/`. I see we are already using it for some of the logging in this file, could we update to make sure we’re using that and not seelog directly in `ecs-agent/` (whether that be in this PR or a future one)?

File: ecs-agent/wsclient/client.go
Comment (danehlim) [inline ecs-agent/wsclient/client.go:19]: Periodic timeout/disconnection of WebSocket connections seems to be missing from this PR, let’s make sure to add that in if we can.

Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go:19]: Currently, the timeouts are initiated by individual components using the library, i.e. ACS and TACS have their own implementation of timeouts. I had not accounted for that as part of wsclient library, hence it's missing from this PR. Is that something we want to unify?

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go:167]: Thank you for pointing out. Will fix this.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go:231]: Yes, ideally I want to get rid of the check. I had a discussion with multiple folks and able to resolve few other such checks. I am preparing to doc to deal with these two checks. Once everyone agrees on the strategy, we can remove these checks as well. I had raised the PR so that ACS is unblocked, as driving consensus is taking some time.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go:157]: Sure. I will update it.

File: agent/wsclient/client.go
Comment (RichaGangwar) [reply agent/wsclient/client.go:38]: I do not think it makes sense to switch just wsconn imports to ecs-agent/wsclient. The plan is to plug in the ecs-agent/wsclient with ecs-agent/tcs and ecs-agent/acs, and once we have the confidence in testing, we would completely remove the agent/wsclient package. It made sense to move the utils to avoid the cyclic dependency issue between agent and ecs-agent.

File: ecs-agent/wsclient/client.go
Comment (danehlim) [reply ecs-agent/wsclient/client.go:231]: Sounds good, will look forward to this in a followup PR then.

File: ecs-agent/wsclient/client.go
Comment (danehlim) [reply ecs-agent/wsclient/client.go:19]: Yeah, I believe we decided that we want to have the same logic for refreshing our websocket connection across all our websocket connections (i.e., ACS and TACS websocket connections). I'll reach out to you offline and point you in the direction of these previous discussions.

File: agent/wsclient/client.go
Comment (danehlim) [reply agent/wsclient/client.go:38]: I was just thinking since the files in `agent/wsclient/wsconn` and `ecs-agent/wsclient/wsconn` are identical, we could just have a single one living in `ecs-agent/`to refer to as a source of truth since we will be removing `agent/wsclient/wsconn` in the future anyway.

But I don't feel super strongly about this, your plan achieves the same end result and works for me.

File: ecs-agent/utils/sign_http_request.go
Comment (danehlim) [inline ecs-agent/utils/sign_http_request.go:32]: This should be updated to use logger instead of seelog as well (per [this previous comment](https://github.com/aws/amazon-ecs-agent/pull/3690#discussion_r1192760046)).

File: ecs-agent/wsclient/client.go
Comment (ohsoo) [inline ecs-agent/wsclient/client.go:72]: Can we move the [exitcodes package](https://github.com/aws/amazon-ecs-agent/blob/883211e24d9255023a68f6351001c291af8a699b/agent/sighandlers/exitcodes/exitcodes.go) to /ecs-agent? Perhaps in a fast follow-up PR, if not this one.

File: ecs-agent/wsclient/client.go
Comment (ohsoo) [inline ecs-agent/wsclient/client.go:130]: nit: typo, RequestResonder -> RequestResponder

File: ecs-agent/wsclient/client.go
Comment (ohsoo) [inline ecs-agent/wsclient/client.go:166]: Does Cfg need to be exported?

File: ecs-agent/wsclient/client.go
Comment (ohsoo) [inline ecs-agent/wsclient/client.go:153]: nit: "WSClientMinAgentConfig is a subset[...]"

File: ecs-agent/wsclient/client.go
Comment (ohsoo) [inline ecs-agent/wsclient/client.go:458]: nit: weird newline

File: ecs-agent/wsclient/client_factory.go
Comment (danehlim) [inline ecs-agent/wsclient/client_factory.go:12]: This needs to be updated to include a `*wsclient.WSClientMinAgentConfig` parameter in the function signature.

File: ecs-agent/wsclient/mock/utils/utils.go
Comment (ohsoo) [inline ecs-agent/wsclient/mock/utils/utils.go:25]: Is this TODO doable within this PR?

File: ecs-agent/utils/sign_http_request.go
Comment (RichaGangwar) [reply ecs-agent/utils/sign_http_request.go]: Fixed. Thank you for pointing out.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go:72]: Sure. Will move it as a follow-up PR. This is already quite bulky. But thank you for pointing out.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go]: Fixed. Thanks.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go:164]: Yes, it would need to be populated by the ACS/TACS libraries. So it needs to be exported.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go]: Fixed. Thanks.

File: ecs-agent/wsclient/client.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client.go]: Fixed.

File: ecs-agent/wsclient/mock/utils/utils.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/mock/utils/utils.go:25]: This mockServer is heavily used in all existing test cases. Maybe it can be picked up in a separate PR. But it would need good amount of rewrite.

File: ecs-agent/wsclient/client_test.go
Comment (danehlim) [inline ecs-agent/wsclient/client_test.go:352]: Are there missing tests here on L349-L352, or what are these comments?

Comment (RichaGangwar) [reply ecs-agent/wsclient/client_test.go:352]: Missed removing the comments. I was using it while writing new tests to ensure the coverage requirements are met. Will remove them in latest commit.

File: ecs-agent/wsclient/client_test.go
Comment (ohsoo) [inline ecs-agent/wsclient/client_test.go:107]: Can we change this name to be `dummyClientServer` or `testClientServer` or similar? That seems a lot more clear to me.

File: ecs-agent/wsclient/client_test.go
Comment (ohsoo) [inline ecs-agent/wsclient/client_test.go:49]: Lets update comments to end with periods/have proper punctuation if they don't already

**Review by ohsoo — COMMENTED** [review]
Let's do some commit squashing please. Overall, not seeing any major concerns!

File: ecs-agent/wsclient/client_test.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client_test.go]: Sure. Done.

File: ecs-agent/wsclient/client_test.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client_test.go]: Fixed.

File: ecs-agent/wsclient/client_test.go
Comment (RichaGangwar) [reply ecs-agent/wsclient/client_test.go]: Sure. Fixed.

**Comment by RichaGangwar** [comment]
> Let's do some commit squashing please. Overall, not seeing any major concerns!

Sure Soo.

File: ecs-agent/wsclient/client_test.go
Comment (danehlim) [inline ecs-agent/wsclient/client_test.go:423]: nit (not blocking, can be added in a followup PR so that this PR can get merged and unblock other work ASAP): 

Can we use `context.Canceled.Error()` here instead of hardcoding "context canceled" (i.e., similar to what we do on L283).

**Review by danehlim — DISMISSED** [review]
Approved under the agreement that we will fast follow with a PR to include periodic timeout/disconnection of websocket connections logic as part of the wsclient itself (per [this comment thread](https://github.com/aws/amazon-ecs-agent/pull/3690#discussion_r1192763621)) and that we will squash commits before merging.

**Review by danehlim — APPROVED** [review]

**Review by Realmonia — APPROVED** [review]

**Review by fierlion — APPROVED** [review]
Can you update the description to highlight TODOs from code and from comments so we can track these in our release process?

**Pull request merged**
